### DISCUSSION TITLE
Fetching an url that redirects with a (stream) body

### DISCUSSION THREAD

**Discussion by feugy** [nodejs/undici#1618] [Q&A] [answered]
Hello Undici community!

I'd like to understand the behavior in a very specific corner case. Maybe it's intended.

I have this [endpoint](https://responseheader-logging-feugy-damien-s-test-team.vercel.app/api/redirect) which returns a 302 redirection to another route, **with a stream body containing a0a1a2a3a4a5**.

When fetched with curl the body is displayed.
<details>
<summary> curl command & results </summary>
<p>

```
curl -i https://responseheader-logging-feugy-damien-s-test-team.vercel.app/api/redirect
HTTP/2 302
cache-control: public, max-age=0, must-revalidate
date: Fri, 26 Aug 2022 14:48:02 GMT
location: /api/route
server: Vercel
strict-transport-security: max-age=63072000; includeSubDomains; preload
x-robots-tag: noindex
x-vercel-cache: MISS
x-vercel-id: cdg1::6thck-1661525281810-22290825160c

a0a1a2a3a4a5%
```

</p>
</details>

When fetched with Undici, the request is aborted and fails if one attempts to consume the body ([reproduction](https://replit.com/@poiupoiuy/fetch-redirect-stream#index.js)), as opposed to node-fetch.

The [spec](https://fetch.spec.whatwg.org/#http-fetch) (step #8)) provides guidance about redirection with body over http2, but also states that the client should return the response.

What's your take?
Thanks a million.

<details>
<summary>Here's the endpoint code, for the curious</summary>
<p>

```js
// running on vercel, edge function
export default function handle(request) {
  const location = new URL('/api/route', request.url)
  const stream = new ReadableStream({
    start(controller) {
      let count = 0
      const l = setInterval(() => {
        controller.enqueue('a' + count++)
        if (count > 5) {
          controller.close()
          clearInterval(l)
        }
      }, 100)
    },
  })
  return new Response(stream, { status: 302, headers: { location } })
}

export const config = { runtime: 'experimental-edge' }
```

</p>
</details>

**Comment by mcollina** [comment]
cc @KhafraDev could you take a look?

**Comment by KhafraDev** [comment]
Yes, this appears to be a mistake in undici (step 8.1, as mentioned).

The connection is destroyed here:
https://github.com/nodejs/undici/blob/6a87bfb38b3f4a28be81d2cc44a80083a0e4f798/lib/fetch/index.js#L1051

Which then aborts the request here:
https://github.com/nodejs/undici/blob/6a87bfb38b3f4a28be81d2cc44a80083a0e4f798/lib/fetch/index.js#L1553
which you can then follow the stack trace, but is relatively unimportant.

---

The connection *should not* be destroyed because undici doesn't support http/2.

You could submit a PR to remove this line with a reproducible example or I can (a little busy atm).

**Reply by ronag** [reply]
The line should only be removed for the "manual" case which we handle differently from the spec.

**Reply by KhafraDev** [reply]
It doesn't appear as any of the checks in step 1 are done and http/2 is unsupported so there's no way to check if a connection uses it. Maybe we can add a utility function that gets implemented once http/2 is supported to add the checks in.

```
If actualResponse’s status is not 303, request’s body is not null, and the connection uses HTTP/2,
then user agents may, and are even encouraged to, transmit an RST_STREAM frame.
```

So something like this maybe?
```diff
diff --git a/lib/fetch/index.js b/lib/fetch/index.js
index c7c88ec..dc247e1 100644
--- a/lib/fetch/index.js
+++ b/lib/fetch/index.js
@@ -1048,7 +1048,10 @@ async function httpFetch (fetchParams) {
     // and the connection uses HTTP/2, then user agents may, and are even
     // encouraged to, transmit an RST_STREAM frame.
     // See, https://github.com/whatwg/fetch/issues/1288
-    fetchParams.controller.connection.destroy()
+    const connection = fetchParams.controller.connection
+    if (actualResponse.status !== 303 && request.body != null && connectionUsesHttp2(connection)) {
+      connection.destroy()
+    }

     // 2. Switch on request’s redirect mode:
     if (request.redirect === 'error') {
```

We would also have to ensure that aborting the connection is the same as emitting an RST_STREAM frame.

**Reply by ronag** [reply]
Don't get hung up on that it says http2. The logic is still the same regardless, i.e. the connection should be closed when following the spec. The only reason this is a problem is because we allow manual redirect handling, which the spec doesn't allow. So it's that path that needs to behave differently.

**Reply by ronag** [reply]
```
if (request.redirect !== 'manual') {
    connection.destroy()
}
```

**Reply by KhafraDev** [reply]
I understand what you're saying now, makes sense.

**Reply by feugy** [reply]
Thanks for jumping on the question folks.
Trying @ronag's suggestion on this particular case works.
What other cases is it breacking @KhafraDev?

I just wrapped the original `fetchParams.controller.connection.destroy()` into an if
```js
  if (redirectStatus.includes(actualResponse.status)) {
    // 1. If actualResponse’s status is not 303, request’s body is not null,
    // and the connection uses HTTP/2, then user agents may, and are even
    // encouraged to, transmit an RST_STREAM frame.
    // See, https://github.com/whatwg/fetch/issues/1288
    if (request.redirect !== 'manual') {
      fetchParams.controller.connection.destroy()
    } 
```

**Reply by KhafraDev** [reply]
It works, I was writing a test which was inherently wrong which caused the issue 😄.

**Reply by feugy** [reply]
Oh, great!
Do you have something ready, or should I open a PR?

**Reply by KhafraDev** [reply]
I don't have much, feel free to open a pr

**Comment by feugy** [comment, marked as answer]
Here's the PR: https://github.com/nodejs/undici/pull/1627
